### PR TITLE
Implement file space strategies.

### PR DESCRIPTION
**Pull request by 1uc**
Implements support for the different file space strategies:
https://portal.hdfgroup.org/display/HDF5/H5P_SET_FILE_SPACE_STRATEGY
https://portal.hdfgroup.org/display/HDF5/H5P_SET_FILE_SPACE_PAGE_SIZE

including paged allocation with configurable page size. Since both are "file creation properties". This MR also includes a new ctor for `File` which enables creating files specifying both creation and access properties.

### CODE REVIEW

File: include/highfive/H5File.hpp [hunk 1]
```diff
@@ -55,6 +55,18 @@ class File: public Object, public NodeTraits<File>, public AnnotateTraits<File>
...
+    /// \param filename: filepath of the HDF5 file
+    /// \param openFlags: Open mode / flags ( ReadOnly, ReadWrite)
+    /// \param fileAccessProps: the file access properties
+    ///
+    /// Open or create a new HDF5 file
+    File(const std::string& filename,
+         unsigned openFlags,
+         const FileCreateProps& fileCreateProps,
+         const FileAccessProps& fileAccessProps = FileAccessProps::Default());
+
```
Comment (alkino) [inline include/highfive/H5File.hpp:69]: SHould you instead add a new Prop to the ctor above?

Comment (1uc) [reply include/highfive/H5File.hpp:69]: I'm worried about code such as:
```
auto fapl = FileAccessProps{};
File("foo.h5", File::ReadOnly, fapl);
```
if we simply change the signature to:
```
File(const std::string& filename,
     unsigned openFlags = ...,
     const FileCreateProps& fileCreateProps = FileCreateProps::Default(),
     const FileAccessProps& fileAccessProps = FileAccessProps::Default());
```
then that will break. OTOH, if we do:
```
File(const std::string& filename,
     unsigned openFlags = ...,
     const FileAccessProps& fileAccessProps = FileAccessProps::Default(),
     const FileCreateProps& fileCreateProps = FileCreateProps::Default());
```
then the order is different than the one for `H5Fcreate`.

**Comment by 1uc** [comment]
One thing to consider is if we want to start adding more and more methods to `File`, e.g. `File::getFileSpacePageSize()`. Instead one could pursue a strategy that allows reading/extracting values from property lists.

File: include/highfive/H5File.hpp (hunk 1 above)
Comment (matz-e) [reply include/highfive/H5File.hpp:69]: `createDataSet` also has first create, then access properties. So this works for me the best in terms of consistency.

File: include/highfive/H5PropertyList.hpp [hunk 2]
```diff
@@ -194,6 +194,70 @@ class MetadataBlockSize {
...
+///
+/// See the upstream documentation of `H5Pget_file_space_strategy` for more details. Essentially,
+/// it enables configuring how space is allocate in the file.
+///
+class FileSpaceStrategy {
+  public:
+    ///
+    /// \brief Create a file space strategy property.
+    ///
+    /// \param strategy  The HDF5 free space strategy.
```
Comment (matz-e) [inline include/highfive/H5PropertyList.hpp:207]:
```suggestion
    /// \param strategy The HDF5 free space strategy.
```
Whitespace, please.

File: include/highfive/bits/H5File_misc.hpp [hunk 3]
```diff
@@ -98,14 +105,63 @@ inline hsize_t File::getMetadataBlockSize() const {
...
+    return strategy;
+}
+
+inline hsize_t File::getFileSpacePageSize() const {
+    auto fcpl = getCreatePList();
+    hsize_t page_size;
+
+    if (getFileSpaceStrategy() != H5F_FSPACE_STRATEGY_PAGE) {
+        HDF5ErrMapper::ToException<FileException>(
+            std::string("Cannot obtain page size, because not using paged allocation."));
```
Comment (matz-e) [inline include/highfive/bits/H5File_misc.hpp:136]:
```suggestion
            std::string("Cannot obtain page size as paged allocation is not used."));
```

**Review by matz-e — DISMISSED** [review]

File: include/highfive/H5PropertyList.hpp (hunk 2 above)
Comment (matz-e) [reply include/highfive/H5PropertyList.hpp:208]: Still there? Is this from formatting? The line does not align, either. I'm puzzled.

File: include/highfive/H5PropertyList.hpp [hunk 4]
```diff
@@ -193,7 +193,59 @@ class MetadataBlockSize {
...
+    /// \param perist Should free space managers be persisted across file closing and reopening.
```
Comment (matz-e) [inline include/highfive/H5PropertyList.hpp:209]:
```suggestion
    /// \param persist Should free space managers be persisted across file closing and reopening.
```

File: include/highfive/bits/H5File_misc.hpp (hunk 3 above)
Comment (matz-e) [reply include/highfive/bits/H5File_misc.hpp:138]: `because not using` is missing a subject and feels like a kludge.

**Review by matz-e — APPROVED** [review]

**Pull request merged**